New blog for VMIM RBAC hardening (#981)
kubevirt/kubevirt.github.io: _posts/2025-02-26-Hardening-VMIM.md

---
layout: post
author: tiraboschi
title: VirtualMachineInstanceMigrations RBAC hardening
description: Apply the principle of least privilege (PoLP) to VirtualMachineInstanceMigrations
navbar_active: Blogs
pub-date: February 26
pub-year: 2025
category: news
tags:
  [
    "VMIM",
    "migrate",
    "migrations",
    "RBAC",
    "hardening",

>     "security"

Reviewer: ```suggestion
    "security",
    "v1.5"
```

  ]
comments: true
---

### Context

> The request to live migrate a VM is represented by an instance of the VirtualMachineInstanceMigration kind.

Reviewer: ```suggestion
The request to live migrate a VM is represented by a `VirtualMachineInstanceMigration` instance.
```

> VirtualMachineInstanceMigration is a namespaced CRD, and its instances are expected to be in the namespace of the VM they refer to.

Reviewer: ```suggestion
A VirtualMachineInstanceMigration (VMIM) is a namespaced CRD, and its instances are expected to be in the namespace of the VM they refer to.
```

> Up to KubeVirt 1.4, by default, a namespace admin (usually a namespace "owner" in a less formal definition) was able to create VMs and also VirtualMachineInstanceMigration (VMIM) objects to enqueue a live migration request for a VM within their namespace.

Reviewer: ```suggestion

Up to KubeVirt 1.4, by default, a namespace admin (usually a namespace "owner" in a less formal definition) was able to create VMs and also VMIM objects to enqueue a live migration request for a VM within their namespace.
```

> On the other hand, live migration is also a building block for infrastructure-critical operations like node drains or upgrades. Therefore, any namespace admin allowed to create or delete VMIM objects could delay or even prevent cluster-critical operations such as an upgrade or a node drain.

Reviewer: ```suggestion
On the other hand, live migration is also a building block for infrastructure-critical operations like node drains or upgrades. Therefore, any namespace admin was allowed to create or delete VMIM objects, which could delay or even prevent cluster-critical operations such as an upgrade or a node drain.
```

> A malicious, less-privileged user could abuse this, causing a kind of DoS at the cluster level. Even worse, Kubernetes RBAC permissions are purely additive (there are no "deny" rules), and KubeVirt roles are constantly reconciled by the virt-operator, so even a conscious cluster admin was unable to deny these permissions as a precautionary measure.

Reviewer: ```suggestion

A malicious, less-privileged user could abuse this, causing a kind of DoS at the cluster level. Even worse, Kubernetes RBAC permissions are purely additive (there are no "deny" rules), and KubeVirt roles are constantly reconciled by the virt-operator, so even a cluster admin who was aware of the issue was unable to deny these permissions as a precautionary measure.
```

For this reason, starting from KubeVirt 1.5, create/delete/update rights will no longer be granted by default to all namespace admins, in accordance with the principle of least privilege (PoLP).
A new convenience ClusterRole named `kubevirt.io:migrate` has been introduced to allow cluster admins to easily grant this permission to selected users.

### Side effects on hotplug operations
Device hotplug operations, at least for CPU and memory, implicitly trigger a live migration executed by the virt-controller on behalf of the user. These operations will not be affected by this change. Under some circumstances or cluster configurations, live migrations are not automatically triggered when NIC devices are hotplugged. In such cases, the only option for namespace admins is to request VMIM permissions from a cluster admin to manually trigger the migration or concatenate two device hotplug operations (where the second one will implicitly complete the NIC hotplug).

> ### Cluster-admins tasks

Reviewer: ```suggestion
### Cluster-admin tasks
```

---
layout: post
author: tiraboschi
title: VirtualMachineInstanceMigrations RBAC hardening
description: Apply the principle of least privilege (PoLP) to VirtualMachineInstanceMigrations
navbar_active: Blogs
pub-date: February 26
pub-year: 2025
category: news
tags:
  [
    "VMIM",
    "migrate",
    "migrations",
    "RBAC",
    "hardening",
    "security",
    "v1.5"
  ]
comments: true
---

### Context
The request to live migrate a VM is represented by a `VirtualMachineInstanceMigration` instance.
A VirtualMachineInstanceMigration (VMIM) is a namespaced CRD, and its instances are expected to be in the namespace of the VM they refer to.

> Up to KubeVirt 1.4, by default, a namespace admin (usually a namespace "owner" in a less formal definition) was able to create VMs and also VMIM objects to enqueue a live migration request for a VM within their namespace.

Reviewer 2: nit: KubeVirt v1.4

Reviewer 2: Also would this be enough:

a namespace admin/owner 

I am not sure if the explanation is needed.

Author: `namespace owner` is technically wrong but on the other side talking with other users/developers I think it's pretty used.
If we want to be formale e should probably say `a user with the admin role for a specific namespace`.

Reviewer 2: Ok thats fair. "a namespace admin" is probably good enough - I am ok with the existing wording.

> At the same time, live migrations can be triggered as part of infrastructure-critical operations like node drains or upgrades, which are the domain of cluster admins. <br>

Reviewer 2: Is the line break  required?

Author: I think that @aburdenthehand suggested to add it for readability purposes.

> At the same time, live migrations can be triggered as part of infrastructure-critical operations like node drains or upgrades, which are the domain of cluster admins. <br>

Reviewer 2: ```suggestion
At the same time, live migrations can be triggered as part of critical infrastructure operations like node drains or upgrades, which are the domain of cluster admins. 
```

Author: done

> Up to KubeVirt 1.4, by default, a namespace admin (usually a namespace "owner" in a less formal definition) was able to create VMs and also VMIM objects to enqueue a live migration request for a VM within their namespace.
> 
> At the same time, live migrations can be triggered as part of infrastructure-critical operations like node drains or upgrades, which are the domain of cluster admins.
> So, if namespace admins can continously enqueue migration requests or delete scheduled VMIM objects needed for ongoing infrastructure-critical operations, they could delay or even prevent cluster-critical operations started by cluster admins, a role with greater privileges.

Reviewer: ```suggestion
Up to KubeVirt 1.4, by default, a namespace admin (usually a namespace "owner" in a less formal definition) was able to create VMs and also VMIM objects to enqueue a live migration request for a VM within their namespace.
At the same time, live migrations can be triggered as part of infrastructure-critical operations like node drains or upgrades, which are the domain of cluster admins. <br>
So, if namespace admins can continuously enqueue migration requests or delete scheduled VMIM objects needed for ongoing infrastructure-critical operations, they could delay or even prevent cluster-critical operations started by cluster admins, a role with greater privileges.
```
Let's keep this all in the one para, but linebreak for the resulting behaviour to help it stand out. 
Other change, since the newline makes github lose track of changes, typo fix on "continuously".

> So a malicious, less-privileged user could abuse this, causing a kind of DoS at the cluster level.

Reviewer: ```suggestion

It was therefore possible that a malicious, less-privileged user could abuse this, causing a kind of DoS at the cluster level.
```

> It was therefore possible that a malicious, less-privileged user could abuse this, causing a kind of DoS at the cluster level.

Reviewer 2: lesser-privileged ?

Author: done

Even worse, Kubernetes RBAC permissions are purely additive (there are no "deny" rules), and KubeVirt roles are constantly reconciled by the virt-operator, so even a cluster admin who was aware of the issue was unable to deny these permissions as a precautionary measure.

> For this reason, starting from KubeVirt 1.5, create/delete/update rights will no longer be granted by default to all namespace admins, in accordance with the principle of least privilege (PoLP).

Reviewer 2: again nit:
 
KubeVirt v1.5

Reviewer 2: I'm not sure of the value of including the "(PoLP)" as its been spelt out in full before that?

Author: done

> A new convenience ClusterRole named `kubevirt.io:migrate` has been introduced to allow cluster admins to easily grant this permission to selected users.

Reviewer 2: ```suggestion
A new convenient ClusterRole named `kubevirt.io:migrate` has been introduced to allow cluster admins to easily grant this permission to selected users.
```

Author: done
